arch/xtensa/src/esp32/esp32_wlan.c

```diff
@@ -0,0 +1,1589 @@
+ *   packets ready to send.  This is a callback from devif_poll().
+ *   devif_poll() may be called:
+ *
+ *   1. When the preceding TX packet send is complete,
+ *   2. When the preceding TX packet send times out and the interface is
+ *      reset
+ *   3. During normal TX polling
+ *
+ * Input Parameters:
+ *   dev - Reference to the NuttX driver state structure
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ * Assumptions:
+ *   May or may not be called from an interrupt handler.  In either case,
+ *   the network is locked.
+ *
+ ****************************************************************************/
+
+static int esp_txpoll(FAR struct net_driver_s *dev)
+{
+  FAR struct esp_dev_s *priv = (FAR struct esp_dev_s *)dev->d_private;
+
+  DEBUGASSERT(priv->esp_dev.d_buf != NULL);
+
+  /* If the polling resulted in data that should be sent out on the network,
+   * the field d_len is set to a value > 0.
+   */
+
+  if (priv->esp_dev.d_len > 0)
+    {
+      /* Look up the destination MAC address and add it to the Ethernet
+       * header.
+       */
+
+#ifdef CONFIG_NET_IPv4
+#ifdef CONFIG_NET_IPv6
+      if (IFF_IS_IPv4(priv->esp_dev.d_flags))
+#endif
+        {
+          arp_out(&priv->esp_dev);
+        }
+#endif /* CONFIG_NET_IPv4 */
+
+#ifdef CONFIG_NET_IPv6
+#ifdef CONFIG_NET_IPv4
+      else
+#endif
+        {
+          neighbor_out(&priv->esp_dev);
+        }
+#endif /* CONFIG_NET_IPv6 */
+
+      if (!devif_loopback(&priv->esp_dev))
+        {
+          /* Send the packet */
+
+          int ret = esp_transmit(priv);
+          if (ret != OK)
+            {
+              wlerr("TX failed\r\n");
+              return -EBUSY;
+            }
+        }
+    }
+
+  /* If zero is returned, the polling will continue until
+   * all connections have been examined.
+   */
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_rxpoll
+ *
+ * Description:
+ *   Process RX frames
+ *
+ * Input Parameters:
+ *   arg - context of device to use
+ *
+ * Returned Value:
+ *   OK on success
+ *
+ * Assumptions:
+ *   The network is locked.
+ *
+ ****************************************************************************/
+
+static void esp_rxpoll(FAR void *arg)
+{
+  FAR struct esp_dev_s *priv = (FAR struct esp_dev_s *)arg;
+
+  if (priv->esp_dev.d_buf == NULL)
+    {
+      priv->esp_dev.d_buf = priv->rxbuf;
+      priv->esp_dev.d_len = priv->rx_len;
+    }
+  else
+    {
+      wlinfo("priv->esp_dev.d_buf != NULL");
+      return;
+    }
+
+  /* Lock the network and serialize driver operations if necessary.
+   * NOTE: Serialization is only required in the case where the driver work
+   * is performed on an LP worker thread and where more than one LP worker
+   * thread has been configured.
+   */
+
+  esp_takesem(priv);
+
+  net_lock();
+
+  esp_receive(priv);
+
+  /* Check if a packet transmission just completed.  If so, call esp_txdone.
```

Comment:
esp_txdone does not seem to exist.
Please review this comment.

Reply:
Due to some reason, the TX done function will be added in next MR, because this function should be supported by wifi internal driver.